### PR TITLE
WinSCard: smart card emulation implementation

### PR DESCRIPTION
**Pull request by TheBestTvarynka**
Hi,
In this PR I've implemented the smart card emulation in the separate winscard-rs crate. Changes in this PR do not affect credentials handling, ffi, and other critical parts of the library.
Overall, this is the first pull request in a series of pull requests related to winscard implementation and integration. I create a single pull request to merge the winscard-rs-impl branch into the master only when all such pull requests will be reviewed and merged.

As the main specification for emulation, I took the open specification for PIV smart cards.

Docs & references:
* [Interfaces for Personal Identity Verification – Part 1: PIV Card Application Namespace, Data Model and Representation](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf)
* [Teleport's rdpclient](https://github.com/gravitational/teleport/tree/46c1cab7c1b59317722f0946e643bdb387f8964d/lib/srv/desktop/rdp/rdpclient)

### CODE REVIEW

File: crates/winscard-rs/src/scard.rs [hunk 1]
```diff
@@ -0,0 +1,783 @@
...
+            Value::Constructed(_) => {
+                // this tag must contain a primitive value
+                return Err(Error::new(
+                    ErrorKind::InvalidValue,
+                    "TLV structure is invalid: challenge field contains constructed value".to_string(),
+                ));
+            }
+        };
+        // TODO: actually sign the challenge
+        // let signed_challenge = sign_hashed_rsa(&self.auth_pk, challenge)?;
```
Comment (TheBestTvarynka) [inline crates/winscard-rs/src/scard.rs:330]: This code was one of the problem places I found during debugging. I have a fix for it. I'll provide a separate pull request with all fixes

File: ffi/Cargo.toml [hunk 2]
```diff
@@ -30,5 +30,5 @@ tracing-subscriber = { version = "0.3", features = ["std", "fmt", "local-time",
 
 [target.'cfg(windows)'.dependencies]
 symbol-rename-macro = { path = "./symbol-rename-macro" }
-winapi = "0.3" # FIXME: replace by windows / windows-sys crates
+winapi = { version =  "0.3", features = ["winerror"] } # FIXME: replace by windows / windows-sys crates
```
Comment (CBenoit) [inline ffi/Cargo.toml:33]: question: Is it a leftover, or something necessary?

File: crates/winscard-rs/src/scard.rs [hunk 3]
```diff
@@ -0,0 +1,783 @@
...
+            ]
+        }
+
+        prop_compose! {
+            fn arb_response()(status in arb_status(), data in option::of(collection::vec(any::<u8>(), 0..256))) -> Response {
+                Response::new(status, data)
+            }
+        }
+
+        proptest! {
```
Comment (CBenoit) [inline crates/winscard-rs/src/scard.rs:403]: praise: I’m happy to see proptest tests :100:

File: crates/winscard-rs/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "winscard-rs"
```
Comment (CBenoit) [inline crates/winscard-rs/Cargo.toml:2]: suggestion: Let’s call the crate `winscard` without the `-rs` suffix. The `-rs` suffix is nice to have in the name of repositories to differentiate from other popular crates with similar names, but we already know that all crates are Rust code.
https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case

File: ffi/Cargo.toml (hunk 2 above)
Comment (TheBestTvarynka) [reply ffi/Cargo.toml:33]: we need `winapi::shared::winerror::ERROR_SUCCESS` for _tsssp_ but `winerror` module is behind the _winerror_ feature

File: crates/winscard-rs/src/scard.rs [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+    use rand::distributions::Uniform;
+    use rand::Rng;
+    use rsa::traits::PublicKeyParts;
+    use rsa::BigUint;
+
+    pub use super::*;
+    use crate::ber_tlv::ber_tlv_length_encoding;
+
+    #[cfg(feature = "proptest")]
+    mod proptests {
```
[the remark this answers is not in the document]
Comment (TheBestTvarynka) [reply crates/winscard-rs/src/scard.rs]: > Do we really need to put the proptest tests behind a "proptest" feature gate

it's a leftover 😅. I just tested the [`proptest-derive`](https://crates.io/crates/proptest-derive/0.4.0) macro but then found out that it is currently experimental.
I already pushed the fix

File: crates/winscard-rs/Cargo.toml (hunk 4 above)
Comment (TheBestTvarynka) [reply crates/winscard-rs/Cargo.toml]: done

File: crates/winscard-rs/src/scard.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (TheBestTvarynka) [reply crates/winscard-rs/src/scard.rs]: Also, maybe I should create a separate crate for prop strategies like in the IronRDP. But I think it's ok for now to place them here

[the remark this answers is not in the document]
Comment (CBenoit) [reply crates/winscard-rs/src/scard.rs]: I think so too!

**Review by CBenoit — APPROVED** [review]
LGTM! There are conflicts to fix, and we can merge. I’ll either rebase and merge if the history is cleaned (descriptive commit individually compiling) or squash in a single commit. (It may be simpler to squash in this situation.)

**Comment by TheBestTvarynka** [comment]
We can squash all intermediate PRs (like this one) but rebase and merge the final PR into the master.

**Comment by CBenoit** [comment]
Sounds reasonable to me!!

**Pull request merged**